src/neural/blas/encoder.h

```diff
@@ -0,0 +1,122 @@
+#include <cstddef>
+#include <cmath>
+
+namespace lczero {
+
+namespace {
+
+template <typename T>
+using EigenMatrixMap =
+    Eigen::Map<Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>>;
+
+template <typename T>
+using ConstEigenMatrixMap =
+    Eigen::Map<const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>>;
+
+}
+
+void LayerNorm2DWithSkipConnection(const size_t batch_size, const size_t channels,
+                                   float* data, const float* skip, const float* gammas,
+                                   const float* betas, float epsilon)
+{
+  for (size_t i = 0; i < batch_size; i++) {
+    // Sum in dimension C.
+    float mean = 0;
+    for (size_t c = 0; c < channels; ++c) {
+      data[i * channels + c] += skip[i * channels + c];
+      mean += data[i * channels + c];
+    }
+
+    // Mean.
+    mean /= channels;
+    
+    // Variance
+    float var = 0;
+    for (size_t c = 0; c < channels; ++c) {
+      auto diff = data[i * channels + c] - mean;
+      var += diff * diff;
+    }
+    var /= channels;
+
+    // Norm.
+    for (size_t c = 0; c < channels; ++c) {
+      data[i * channels + c] = betas[c] + gammas[c]
+        * (data[i * channels + c] - mean) / std::sqrt(var + epsilon);
+    }
+
+  }
+
+}
+
+void AttentionMatmul2D(const bool transpose_a, const bool transpose_b,
+                     const size_t batch_size, const size_t M,
+                     const size_t N, const size_t K,
+                     const float scaling,
+                     const float* input1, const float* input2,
+                     float* output, bool use_eigen)
+{
+  for (auto batch = size_t{0}; batch < batch_size; batch++) {
+    const float* A = &input1[batch * M * K];
+    const float* B = &input2[batch * N * K];
+    float* C = &output[batch * M * N];
+    if (use_eigen) {
+      auto C_mat = EigenMatrixMap<float>(C, N, M);
+
+      if (transpose_a && transpose_b) {
+        C_mat.noalias() = scaling
+          * ConstEigenMatrixMap<float>(B, K, N).transpose()
+          * ConstEigenMatrixMap<float>(A, M, K).transpose();
+      }
+      else if (transpose_a) {
+        C_mat.noalias() = scaling
+          * ConstEigenMatrixMap<float>(B, N, K)
+          * ConstEigenMatrixMap<float>(A, M, K).transpose();
+      }
+      else if (transpose_b) {
+        C_mat.noalias() = scaling
+          * ConstEigenMatrixMap<float>(B, K, N).transpose()
+          * ConstEigenMatrixMap<float>(A, K, M);
+      }
+      else {
+        C_mat.noalias() = scaling
+          * ConstEigenMatrixMap<float>(B, N, K)
+          * ConstEigenMatrixMap<float>(A, K, M);
+      }
+    } else {
+#ifdef USE_BLAS
+      cblas_sgemm(CblasRowMajor, transpose_a ? CblasTrans : CblasNoTrans,
+                  transpose_b ? CblasTrans : CblasNoTrans,
+                  M, N, K, scaling, A, transpose_a ? M : K,
+                  B, transpose_b ? K : N, 0.0f, C, N);
+#else
+      // Should never get here.
+      throw Exception("Blas backend internal error");
```

Comment:
This breaks the dx12 build - I think making `use_eigen` a template parameter should fix it.

Reply:
Done